pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go

```diff
@@ -0,0 +1,190 @@
+	controller = NewController(
+		e.Manager.GetClient(),
+		corev1.NewForConfigOrDie(e.Manager.GetConfig()),
+		fake.NewFactory(cloudprovider.Options{}),
+	)
+	e.Manager.Register(controller)
+})
+
+var _ = BeforeSuite(func() {
+	Expect(env.Start()).To(Succeed(), "Failed to start environment")
+})
+
+var _ = AfterSuite(func() {
+	Expect(env.Stop()).To(Succeed(), "Failed to stop environment")
+})
+
+var _ = Describe("Reallocation", func() {
+	var ns *environment.Namespace
+
+	BeforeEach(func() {
+		var err error
+		ns, err = env.NewNamespace()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		ExpectCleanedUp(ns.Client)
+	})
+
+	Context("Reconciliation", func() {
+		It("should label nodes for underutilized", func() {
+			p := &v1alpha1.Provisioner{
+				ObjectMeta: metav1.ObjectMeta{Name: strings.ToLower(randomdata.SillyName()), Namespace: ns.Name},
+				Spec: v1alpha1.ProvisionerSpec{
+					Cluster: &v1alpha1.ClusterSpec{Name: "test-cluster", Endpoint: "http://test-cluster", CABundle: "dGVzdC1jbHVzdGVyCg=="},
+				},
+			}
+			node := test.NodeWith(test.NodeOptions{
+				Labels: map[string]string{
+					v1alpha1.ProvisionerNameLabelKey:      p.Name,
+					v1alpha1.ProvisionerNamespaceLabelKey: p.Namespace,
+				},
+			})
+			ExpectCreatedWithStatus(ns.Client, node)
+
+			ExpectCreated(ns.Client, p)
+			ExpectEventuallyReconciled(ns.Client, p)
+
+			Expect(ns.Client.Get(context.Background(), client.ObjectKey{Name: node.Name}, node)).To(Succeed())
+			Expect(node.Labels).To(HaveKeyWithValue(v1alpha1.ProvisionerPhaseLabel, v1alpha1.ProvisionerUnderutilizedPhase))
+			Expect(node.Annotations).To(HaveKey(v1alpha1.ProvisionerTTLKey))
+		})
+
+		It("should remove labels from utilized nodes", func() {
+			p := &v1alpha1.Provisioner{
+				ObjectMeta: metav1.ObjectMeta{Name: strings.ToLower(randomdata.SillyName()), Namespace: ns.Name},
+				Spec: v1alpha1.ProvisionerSpec{
+					Cluster: &v1alpha1.ClusterSpec{Name: "test-cluster", Endpoint: "http://test-cluster", CABundle: "dGVzdC1jbHVzdGVyCg=="},
+				},
+			}
+			node := test.NodeWith(test.NodeOptions{
+				Labels: map[string]string{
+					v1alpha1.ProvisionerNameLabelKey:      p.Name,
+					v1alpha1.ProvisionerNamespaceLabelKey: p.Namespace,
+					v1alpha1.ProvisionerPhaseLabel:        v1alpha1.ProvisionerUnderutilizedPhase,
+				},
+				Annotations: map[string]string{
+					v1alpha1.ProvisionerTTLKey: time.Now().Add(time.Duration(100) * time.Second).Format(time.RFC3339),
+				},
+			})
+			pod := test.PodWith(test.PodOptions{
+				Name:       strings.ToLower(randomdata.SillyName()),
+				Namespace:  p.Namespace,
+				NodeName:   &node.Name,
+				Conditions: []v1.PodCondition{{Type: v1.PodReady, Status: v1.ConditionTrue}},
+			})
+
+			ExpectCreatedWithStatus(ns.Client, node)
+			ExpectCreatedWithStatus(ns.Client, pod)
+
+			ExpectCreated(ns.Client, p)
+			ExpectEventuallyReconciled(ns.Client, p)
+
+			nodeCheck := &v1.Node{}
```

Comment:
I think you can reuse the node object if you wanted to. Otherwise I might name this updatedNode

Reply:
Since this test requires deleting the labels and annotations, I think there's some weird merge-like behavior happening in the Get function if I reuse node here. Every other test works fine without this, but I'll change this usage to be `updatedNode`